### ISSUE TITLE
Implementation of SemVer does not line up with expectation in semver.js

### ISSUE THREAD

**Issue by mikesol** [purescript/registry-dev#246]
I believe that semver.js is expecting just a version (a string) when it does stuff like comparing, but in the FFI we are passing it a record. I think the fix here would be to pass semver.js the `version` field of the record in `SemVer`. As I'm not exactly sure if this is right, I'll file it as an issue, and if folks confirm this is indeed the case, I can make a PR.

**Comment by thomashoneyman** [comment]
It's a bit confusing -- the `compare` function from `node-semver` is implemented like this:

https://github.com/npm/node-semver/blob/f1e4e293de1ad73fa250336bfcfb9b4142d21d1b/functions/compare.js#L2-L3
```js
const compare = (a, b, loose) =>
  new SemVer(a, loose).compare(new SemVer(b, loose))
```

which implies that `a` and `b` are versions, since they get constructed into `SemVer` classes. But the `SemVer` constructor itself checks whether it's been passed an existing `SemVer`:

https://github.com/npm/node-semver/blob/f1e4e293de1ad73fa250336bfcfb9b4142d21d1b/classes/semver.js#L7-L20
```js
class SemVer {
  constructor (version, options) {
    options = parseOptions(options)


    if (version instanceof SemVer) {
      if (version.loose === !!options.loose &&
          version.includePrerelease === !!options.includePrerelease) {
        return version
      } else {
        version = version.version
      }
    } else if (typeof version !== 'string') {
      throw new TypeError(`Invalid Version: ${version}`)
    }
```

and so, from my reading, passing in a full `SemVer` object is fine.

That said, it's not always clear when some manipulation in PureScript might cause a JS `instanceof` check to no longer be true, and so I would be a little happier passing in just the version string itself. (cc: @f-f)

**Comment by f-f** [comment]
I'm a little confused here - did we uncover a bug? It we're able to write a test that shows funny behaviour then we can do something about it, but if things work fine for all possible testcases then I think the current implementation is alright?

**Comment by thomashoneyman** [comment]
> but if things work fine for all possible testcases

I don't know if I'm bold enough to say _that_, but I do think things are working correctly right now.

@mikesol encountered crashes in the JS layer when constructing the `SemVerJS` record directly (ie. not via `parseSemVer`), but presumably that's because it's failing the `instanceof` check. So first the constructor checks if we passed a `SemVer`, which we didn't (we just passed a bare record), and then it checks if we passed a string (again, no, we passed a bare record), and then crashes.

However, I would be open to using the simple string comparison version because I've encountered funky JS issues before with `instanceof` checks when manipulating JS structures in PS when they're copied (for example, `map`ping over an array of them). My JS-foo isn't good enough to be totally clear about when this causes issues and when it doesn't, so I like to interact with JS apis using the simplest types available.

That way I _do_ feel bold enough to say it satisfies all possible testcases!
